pkg/k8s/completion.go

```diff
@@ -0,0 +1,165 @@
+)
+
+// CommandCompletion generates CLI suggestions from resources in a given cluster
+// It uses a list of arguments and a substring from the CLI to filter suggestions.
+type CommandCompletion struct {
+	k8sAPI    *KubernetesAPI
+	namespace string
+}
+
+// NewCommandCompletion creates a command completion module
+func NewCommandCompletion(
+	k8sAPI *KubernetesAPI,
+	namespace string,
+) *CommandCompletion {
+	return &CommandCompletion{
+		k8sAPI:    k8sAPI,
+		namespace: namespace,
+	}
+}
+
+// Complete accepts a list of arguments and a substring to generate CLI suggestions.
+// `args` represent a list of arguments a user has already enterd in the CLI. These
+// arguments are used for determining what resource type we'd like to receive
+// suggestions for as well as a list of resources names that have already provided.
+// `toComplete` represents the string prefix of a resource name that we'd like to
+// use to search for suggestions
+//
+// If `args` is empty, send back a list of all resource types support by the CLI.
+// If `args` has at least one or more items, assume that the first item in `args`
+// is the resource type we are trying to get suggestions for e.g. Deployment, StatefulSets.
+//
+// Complete is generic enough so that it can find suggestions for any type of resource
+// in a Kubernetes cluster. It does this by first querying what GroupVersion a resource
+// belongs to and then does a dynamic `List` query to get resources under that GroupVersion
+func (c *CommandCompletion) Complete(args []string, toComplete string) ([]string, error) {
+	ctx, cancelFn := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancelFn()
+
+	suggestions := []string{}
+	if len(args) == 0 && toComplete == "" {
+		return StatAllResourceTypes, nil
+	}
+
+	if len(args) == 0 && toComplete != "" {
+		for _, t := range StatAllResourceTypes {
+			if strings.HasPrefix(t, toComplete) {
+				suggestions = append(suggestions, t)
+			}
+		}
+		return suggestions, nil
+	}
+
+	resType, err := CanonicalResourceNameFromFriendlyName(args[0])
+	if err != nil {
+		return nil, fmt.Errorf("%s not a valid resource name", args)
+	}
+
+	// if we are looking for namespace suggestions clear namespace selector
+	if resType == "namespace" {
+		c.namespace = ""
+	}
+
+	// Similar to kubectl, we don't provide resource completion
+	// when the resource provided is in format <kind>/<resourceName>
+	if strings.Contains(args[0], "/") {
+		return []string{}, nil
+	}
+
+	gvr, err := c.getGroupVersionKindForResource(resType)
+	if err != nil {
+		return nil, err
+	}
+
+	uList, err := c.k8sAPI.DynamicClient.
+		Resource(*gvr).
+		Namespace(c.namespace).
+		List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	otherResources := args[1:]
+	for _, u := range uList.Items {
+		name := u.GetName()
+		if strings.HasPrefix(name, toComplete) &&
+			!containsResource(name, otherResources) {
+			suggestions = append(suggestions, name)
+		}
+	}
```

Comment:
I wasn't totally sure what this was for when originally reviewing. It may be worth a comment explaining that this deduplicates resource suggestions so that, for example, if I already have `linkerd-destination` as a resource and want to tab-complete for `linkerd-identity`, it does not resugggest `linkerd-destination`.